### PR TITLE
Replace println where appropriate

### PR DESCRIPTION
**Pull request by flyinprogrammer**
This PR should address: https://github.com/ardanlabs/gotraining/issues/238

### CODE REVIEW

**Comment by flyinprogrammer** [comment]
This is still a WIP - I just figured I'd start to show that real work is being done.

**Comment by flyinprogrammer** [comment]
https://github.com/ardanlabs/gotraining/blob/master/topics/go/concurrency/channels/advanced/example1/example1.go#L30

This actually seems to be the only file that uses the `println` function from the builtin pkg that isn't part of the pointer examples. I'd be more than happy to just submit that as a change, and/or we can all pretend this never happened and ya'll can make the change yourself. I just thought I could help, the more time I've spent on this the more skeptical I am of the value.

**Comment by ardan-bkennedy** [comment]
There is a lot of great fixes here! We need those playground links :(

**Comment by flyinprogrammer** [comment]
these should be all done - i'm going to work on the playground links tonight.

File: topics/go/language/constants/example3/example3.go [hunk 1]
```diff
@@ -14,23 +14,23 @@ func main() {
 		C1 = iota // 2 : Increment by 1
 	)
 
-	fmt.Println("1:", A1, B1, C1)
+	fmt.Printf("1: %d %d %d\n", A1, B1, C1)
```
Comment (jcbwlkr) [inline topics/go/language/constants/example3/example3.go]: Revert the changes in this file. We don't really even run this code we just have it here to talk about the behavior of constants. IMO students will be distracted by the printf verbs here when I want them just focusing on how constants and iota work.

File: topics/go/language/embedding/exercises/exercise1/exercise1.go [hunk 2]
```diff
@@ -53,12 +53,12 @@ func main() {
 	}
 
 	// Display what we are trying to find.
-	fmt.Println("Searching For:", needle)
+	fmt.Printf("Searching For: %s\n", needle)
```
Comment (jcbwlkr) [inline topics/go/language/embedding/exercises/exercise1/exercise1.go]: No behavior change here. Revert to Println

File: topics/go/language/embedding/exercises/exercise1/exercise1.go [hunk 3]
```diff
@@ -53,12 +53,12 @@ func main() {
...
 
 	// Range of each haystack value and check the term.
 	for _, hs := range haystack {
 		if hs.find(needle) {
-			fmt.Printf("FOUND: %+v", hs)
+			fmt.Printf("FOUND: %+v\n", hs)
```
Comment (jcbwlkr) [inline topics/go/language/embedding/exercises/exercise1/exercise1.go]: This was probably a bug before. Thanks for catching.

File: topics/go/language/exporting/exercises/exercise1/exercise1.go [hunk 4]
```diff
@@ -28,8 +28,8 @@ func main() {
 	t.UpdateSold(2)
 
 	// Display each field separately.
-	fmt.Println("Name", t.Name)
-	fmt.Println("Weight", t.Weight)
-	fmt.Println("OnHand", t.OnHand())
-	fmt.Println("Sold", t.Sold())
```
Comment (jcbwlkr) [inline topics/go/language/exporting/exercises/exercise1/exercise1.go]: Capitalize the `n` in `name`

File: topics/go/language/interfaces/advanced/example1/example1.go [hunk 5]
```diff
@@ -22,13 +22,13 @@ type user struct {
 
 // notify implements the notifier interface.
 func (u user) notify() {
-	fmt.Println("Alert", u.name)
+	fmt.Printf("Alert %s\n", u.name)
```
Comment (jcbwlkr) [inline topics/go/language/interfaces/advanced/example1/example1.go]: No behavior change. Revert.

File: topics/go/language/maps/example1/example1.go [hunk 6]
```diff
@@ -25,16 +25,19 @@ func main() {
 	users["Mouse"] = user{"Mickey", "Mouse"}
 	users["Jackson"] = user{"Michael", "Jackson"}
 
+	fmt.Println("\nPrint map")
```
Comment (jcbwlkr) [inline topics/go/language/maps/example1/example1.go]: We don't need the initial `\n` here.

File: topics/go/language/struct_types/example1/example1.go [hunk 7]
```diff
@@ -31,7 +31,10 @@ func main() {
 	}
 
 	// Display the field values.
-	fmt.Println("Flag", e2.flag)
-	fmt.Println("Counter", e2.counter)
-	fmt.Println("Pi", e2.pi)
+	fmt.Printf(
+		"Flag[%t] Counter[%d] Pi[%.6f]\n",
+		e2.flag,
```
Comment (jcbwlkr) [inline topics/go/language/struct_types/example1/example1.go]: Please revert this one. The point in this example is learning how to get/set struct field values using the `.` operator. I think students will be distracted by the printf verbs.

File: topics/go/language/struct_types/example2/example2.go [hunk 8]
```diff
@@ -34,7 +34,10 @@ func main() {
 
 	// Display the values.
 	fmt.Printf("%+v\n", e2)
-	fmt.Println("Flag", e2.flag)
-	fmt.Println("Counter", e2.counter)
-	fmt.Println("Pi", e2.pi)
+	fmt.Printf(
```
Comment (jcbwlkr) [inline topics/go/language/struct_types/example2/example2.go]: Please revert this one. The point in this example is learning how to get/set struct field values using the `.` operator. I think students will be distracted by the printf verbs.

File: topics/go/language/struct_types/example3/example3.go [hunk 9]
```diff
@@ -39,7 +39,10 @@ func main() {
 	// Display the values.
 	fmt.Printf("%+v\n", ex)
 	fmt.Printf("%+v\n", e)
-	fmt.Println("Flag", e.flag)
-	fmt.Println("Counter", e.counter)
-	fmt.Println("Pi", e.pi)
+	fmt.Printf(
+		"Flag[%t] Counter[%d] Pi[%.6f]\n",
+		e.flag,
```
Comment (jcbwlkr) [inline topics/go/language/struct_types/example3/example3.go]: Please revert this one. The point in this example is learning how to get/set struct field values using the `.` operator. I think students will be distracted by the printf verbs.

File: topics/go/language/variables/exercises/exercise1/exercise1.go [hunk 10]
```diff
@@ -19,9 +19,12 @@ func main() {
 	var legal bool
 
 	// Display the value of those variables.
-	fmt.Println(age)
-	fmt.Println(name)
-	fmt.Println(legal)
+	fmt.Printf(
```
Comment (jcbwlkr) [inline topics/go/language/variables/exercises/exercise1/exercise1.go]: Revert all of the changes in this file. This is the first exercise people go through and we are just having them print the values. We want to start small so they are comfortable just building/running a Go program.

File: topics/go/packages/encoding/example2/example2.go [hunk 11]
```diff
@@ -38,9 +38,9 @@ type (
 func main() {
 
 	// Open the file.
-	file, err := os.Open("./topics/go/packages/encoding/example2/data.json")
+	file, err := os.Open("data.json")
```
Comment (jcbwlkr) [inline topics/go/packages/encoding/example2/example2.go]: @ardan-bkennedy do you run this example in your training? And if so do you run it from the root of the training repo or do you run it from this folder? I run it from this folder and always change the path like @flyinprogrammer has done here so I'm fine with the change but I want your input.

File: topics/go/packages/encoding/example3/example3.go [hunk 12]
```diff
@@ -56,21 +56,21 @@ func main() {
...
 	r2, err := json.MarshalIndent(&station, "", "    ")
 	if err != nil {
-		fmt.Println("MarshalIndent", err)
+		fmt.Printf("could not MarshalIndent json: %s\n", err)
 		return
 	}
 
 	// Convert the byte slice to a string and display.
-	fmt.Println(string(r2))
+	fmt.Printf("indented json:\n%s\n", string(r2))
```
Comment (jcbwlkr) [inline topics/go/packages/encoding/example3/example3.go]: We can remove the cast to `string()` if we're using the `%s` verb

File: topics/web/serializers/example1/main.go [hunk 13]
```diff
@@ -26,6 +26,7 @@ func useStreams() {
 		LastName:  "Walker",
 	}
 
+	fmt.Print("json output: ")
```
Comment (jcbwlkr) [inline topics/web/serializers/example1/main.go]: I'd rather not have these labels. Remove this line please

File: topics/web/serializers/example1/main.go [hunk 14]
```diff
@@ -62,7 +63,7 @@ func useBytes() {
 	if err != nil {
 		log.Fatal(err)
 	}
-	fmt.Println(string(b))
```
Comment (jcbwlkr) [inline topics/web/serializers/example1/main.go]: Revert this one please.

File: topics/web/serializers/example4/main.go [hunk 15]
```diff
@@ -54,11 +54,11 @@ func main() {
...
 	}
 
-	print("User with data", b)
+	printMsgData("User with data", b)
 
 	fmt.Println("\nNote in all examples 'email' is missing.")
 }
 
-func print(msg string, data []byte) {
+func printMsgData(msg string, data []byte) {
```
Comment (jcbwlkr) [inline topics/web/serializers/example4/main.go]: This is probably a good change so people don't confuse `print` with a builtin function.

File: main.go [hunk 16]
```diff
@@ -55,7 +55,7 @@ func main() {
 
 	}
 
-	fmt.Println("Listening on port", port)
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply main.go]: I agree

File: topics/go/concurrency/data_race/example5/example5.go [hunk 17]
```diff
@@ -32,5 +32,5 @@ func main() {
 	}()
 
 	wg.Wait()
-	fmt.Println("Final scores:", scores)
+	fmt.Printf("Final scores:\n%+v\n", scores)
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/concurrency/data_race/example5/example5.go]: Agree

File: topics/data/data_versioning/example4/example4.go [hunk 18]
```diff
@@ -33,5 +33,5 @@ func main() {
 	}
 
 	// Output the contents to standard out.
-	fmt.Println(b.String())
+	fmt.Printf("contents:\n%s\n", b.String())
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/data/data_versioning/example4/example4.go]: Agree

File: topics/go/design/composition/assertions/example2/example2.go [hunk 19]
```diff
@@ -47,7 +47,7 @@ func main() {
 		// Perform a type assertion that we have a concrete type
 		// of cloud in the interface value we randomly chose.
 		if v, is := mvs[rn].(cloud); is {
-			fmt.Println("Got Lucky:", v)
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/design/composition/assertions/example2/example2.go]: Agree

File: topics/go/design/composition/assertions/example3/example3.go [hunk 20]
```diff
@@ -26,6 +26,5 @@ func main() {
 	}
 
 	// Display the values.
-	fmt.Println(u)
-	fmt.Println(&u)
+	fmt.Printf("value[%+v], address[%p]\n", u, &u)
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/design/composition/assertions/example3/example3.go]: Agree

File: topics/go/packages/encoding/example2/example2.go (hunk 11 above)
Comment (ardan-bkennedy) [reply topics/go/packages/encoding/example2/example2.go]: I never get to this material.

File: topics/web/serializers/example1/main.go (hunk 13 above)
Comment (ardan-bkennedy) [reply topics/web/serializers/example1/main.go]: Agree

File: topics/web/serializers/example4/main.go (hunk 15 above)
Comment (ardan-bkennedy) [reply topics/web/serializers/example4/main.go]: Agree

File: topics/web/serializers/example1/main.go (hunk 14 above)
Comment (ardan-bkennedy) [reply topics/web/serializers/example1/main.go]: Agree

File: topics/go/design/composition/grouping/example1/example1.go [hunk 21]
```diff
@@ -19,10 +19,11 @@ type Animal struct {
 // Speak provides generic behavior for all animals and
 // how they speak.
 func (a *Animal) Speak() {
-	fmt.Println("UGH!",
-		"My name is", a.Name,
-		", it is", a.IsMammal,
-		"I am a mammal")
+	fmt.Printf(
+		"UGH! My name is %s, it is %s I am a mammal\n",
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/design/composition/grouping/example1/example1.go]: Agree

File: topics/go/design/composition/grouping/example1/example1.go [hunk 22]
```diff
@@ -34,10 +35,12 @@ type Dog struct {
 
 // Speak knows how to speak like a dog.
 func (d *Dog) Speak() {
-	fmt.Println("Woof!",
-		"My name is", d.Name,
-		", it is", d.IsMammal,
-		"I am a mammal with a pack factor of", d.PackFactor)
+	fmt.Printf(
+		"Woof! My name is %s, it is %s I am a mammal with a pack factor of %d.\n",
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/design/composition/grouping/example1/example1.go]: Agree

File: topics/go/design/error_handling/example3/example3.go [hunk 23]
```diff
@@ -20,7 +20,11 @@ type UnmarshalTypeError struct {
 
 // Error implements the error interface.
 func (e *UnmarshalTypeError) Error() string {
-	return "json: cannot unmarshal " + e.Value + " into Go value of type " + e.Type.String()
+	return fmt.Sprintf(
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/design/error_handling/example3/example3.go]: Agree

File: topics/go/design/error_handling/example4/example4.go [hunk 24]
```diff
@@ -1,8 +1,8 @@
 // All material is licensed under the Apache License Version 2.0, January 2004
 // http://www.apache.org/licenses/LICENSE-2.0
 
-// Package example5 provides code to show how to implement behavior as context.
-package example5
+// Package example4 provides code to show how to implement behavior as context.
+package example4
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/design/error_handling/example4/example4.go]: Agree

File: topics/go/exercises/contributors/githubmock/main.go [hunk 25]
```diff
@@ -47,6 +47,6 @@ func main() {
 		log.Fatal(err)
 	}
 
-	fmt.Println("Mock GitHub server listening on", listener.Addr().String())
+	fmt.Printf("Mock GitHub server listening on: %s\n", listener.Addr().String())
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/exercises/contributors/githubmock/main.go]: Agree

File: topics/go/exercises/contributors/part1/main.go [hunk 26]
```diff
@@ -69,6 +69,11 @@ func main() {
 
 	// Loop through the []contributor and print the values.
 	for i, con := range cons {
-		fmt.Println(i, con.Login, con.Contributions)
+		fmt.Printf(
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/exercises/contributors/part1/main.go]: Agree

File: topics/go/exercises/contributors/part2/main.go [hunk 27]
```diff
@@ -41,7 +41,12 @@ func printContributors(repo string, c contributors) int {
 	}
 
 	for i, con := range cons {
-		fmt.Println(i, con.Login, con.Contributions)
+		fmt.Printf(
```
[the remark this answers is not in the document]
Comment (ardan-bkennedy) [reply topics/go/exercises/contributors/part2/main.go]: Agree

File: topics/go/language/constants/example3/example3.go (hunk 1 above)
Comment (ardan-bkennedy) [reply topics/go/language/constants/example3/example3.go]: agree

File: topics/go/language/embedding/exercises/exercise1/exercise1.go (hunk 2 above)
Comment (ardan-bkennedy) [reply topics/go/language/embedding/exercises/exercise1/exercise1.go]: Agree

File: topics/go/language/embedding/exercises/exercise1/exercise1.go (hunk 3 above)
Comment (ardan-bkennedy) [reply topics/go/language/embedding/exercises/exercise1/exercise1.go]: Agree

File: topics/go/language/interfaces/advanced/example1/example1.go (hunk 5 above)
Comment (ardan-bkennedy) [reply topics/go/language/interfaces/advanced/example1/example1.go]: Agree

File: topics/go/language/maps/example1/example1.go (hunk 6 above)
Comment (ardan-bkennedy) [reply topics/go/language/maps/example1/example1.go]: Agree

File: topics/go/language/struct_types/example1/example1.go (hunk 7 above)
Comment (ardan-bkennedy) [reply topics/go/language/struct_types/example1/example1.go]: Agree

File: topics/go/language/struct_types/example3/example3.go (hunk 9 above)
Comment (ardan-bkennedy) [reply topics/go/language/struct_types/example3/example3.go]: Agree

File: topics/go/language/variables/exercises/exercise1/exercise1.go (hunk 10 above)
Comment (ardan-bkennedy) [reply topics/go/language/variables/exercises/exercise1/exercise1.go]: Agree

**Comment by ardan-bkennedy** [comment]
Let's revert everything @jcbwlkr has identified @flyinprogrammer. Again thanks for such an amazing effort to make the material better!

**Comment by flyinprogrammer** [comment]
I'm super sorry for the silence, I've been scrambling to get things together for a super last minute 2 week trip to India. I super appreciate the feedback and will absolutely get all these edits in, with multiple PRs, as soon as I can breath for a minute (hopefully by the end of the week).

**Comment by flyinprogrammer** [comment]
closed in favor of smaller changes: https://github.com/ardanlabs/gotraining/pull/247
